### ISSUE TITLE
Is certificate installation in scope for lego?

### ISSUE THREAD

**Issue by willglynn** [xenolf/lego#126]
`lego` has three verbs – `run`, `renew`, `revoke` – all of which act on certificates and ACME account data stored on the local filesystem within `--path=.lego/`. As of this writing, `lego` makes no attempt to reconfigure web servers, to facilitate transformations of the recorded data (#120), or to otherwise involve itself in what happens after it's obtained a certificate. Once the ACME exchange is complete, `lego` terminates and considers its job done.

If certificate installation is in scope for `lego`, we'd want to define an interface allowing modules to observe the certificate lifecycle, like `GotNewCertificate(cert []byte, key []byte)`. We'd have an [Apache configurator](https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt-apache/letsencrypt_apache/configurator.py) and an [nginx configurator](https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt-nginx/letsencrypt_nginx/configurator.py) like the official Python client. I would also want an implementation that delegates to external processes (shell scripts make excellent glue!), but conceivably we could end up with a module supporting every SSL-capable tool, even those on remote systems. (I want my home router to use a valid certificate!) This rabbit hole goes really deep.

If these problems are not in scope, `lego` stays how it is, existing as an executable version of  the `lego/acme` library. If you need more automation, you should write a script around `lego` – and `lego` should provide enough surface area that it can be used that way (#124, #125). If your needs are too complicated for a shell script, then you should make a new tool importing the `lego/acme` library. Presumably also these tools would not share the `.lego/` directory… unless we want them to, in which case we need some refactoring.

Focusing `lego`'s scope seems to be intentional; @xenolf [wrote](https://github.com/xenolf/lego/issues/111#issuecomment-181670463):

> As far as "plugins" to the CLI go I'm torn back and forth. I intended the CLI to be simple and only a thin wrapper around the library and this would kinda break it.

Thoughts?

**Comment by mholt** [comment]
My instinct tells me that server configuration is outside the scope of lego. I think lego should stay as simple as possible. Its crown jewel is the acme package, which is, IMO, one of the most mature and well-documented ACME clients available today, and it's usable in any Go program. I think of the CLI as more of a simple utility for people with basic needs and don't want any magic, just a simple "get a certificate and put it on disk" program. lego really shines at that as so many other clients have multiple dependencies or do a lot more and are harder to use.

I've started another project, [certs](https://github.com/mholt/certs), which I intend to be a more feature-filled "certificates daemon" (I might even rename it to certsd). I still don't like the idea of web server configuration, but it would at least be more customizable and more capable of bulk certificate management. Essentially a web server but for certificates. Even though that project doesn't have a README yet (I only made the an "issue" function so far), I'd be open to discussing such possibilities over there.

Of course I don't mean to hijack this discussion from lego. If @xenolf wants it here then that's fine. But if not, there's still a venue to have this discussion.

**Comment by willglynn** [comment]
That's pretty much what I was thinking – there's lots of space for tools on top of `lego/acme` and it makes sense for `lego` to stay approximately what it is right now.

I have particular problems involving certificate management on systems that can't directly participate in ACME exchanges. This is why I've been so interested in dark corners of the DNS-01 challenge, why I'm eager to land `ObtainCertificateForCSR()`, and why I'm asking if this kind of stuff belongs here. My gut says this isn't any of `lego`'s business, but I figured it's worth reaching such conclusions together.

**Comment by xenolf** [comment]
Server configuration is indeed a rabbit hole which I don't want this project to venture into. The CLI was built with the intention of being a tight wrapper around the library and involve as little magic as possible.
The main part of this project is the library which I want to be rock solid and a breeze to use.

I think more specialized things are better built in their own tools on top of the library.

**Comment by xenolf** [comment]
> Presumably also these tools would not share the .lego/ directory… unless we want them to, in which case we need some refactoring.

Configuration was intentionally left to the implementing program. People are free to use the CLI as an example, but in no way is this configuration structure set in stone.

**Comment by willglynn** [comment]
Great! Sounds like there's nothing to do here :smiley:
